### ISSUE TITLE
Administrative Notices

### ISSUE THREAD

**Issue by Firstyear** [kanidm/kanidm#1615]
Sometimes when we make changes to Kanidm, in rare occasions we may require administrator intervention. This has happened only once when we changed the format for cron strings, and because the kanidm config is read-only to us, we could not automatically fix this inplace.

In the future we should have a way to supply notifications to administrators so that they are aware of upcoming changes they may need to take actions for. They should also have a way to silence or acknowledge notifications once they have actioned or seen them.

**Comment by Gabgobie** [comment]
I think this can be fixed with a small addition to the docs.

In general, an admin should read the change notes before performing an update. At the same time, the admin should be aware of the existence of said update.

Both can be solved by subscribing to the GitHub releases RSS feed: <https://github.com/kanidm/kanidm/releases.atom>

The only thing missing would be a mention in the docs. This would also alleviate the need for code that checks for configuration from earlier versions or similar and avoids the complexity of a newsletter.

If you want a little more control or disconnect it from the releases, the feed could also be somewhere on <https://kanidm.com>

---

Having this communication directly in Kani would make releases, especially breaking ones, more complex and requires the instance/cluster to keep track of the notifications that are relevant and that the admin has seen (per account?).

Furthermore, the only available place to show admin notifications, if I am not mistaken, is currently the CLI client.

**Comment by Firstyear** [comment]
@Gabgobie I think you may be missing some context here:

This relates to https://kanidm.github.io/kanidm/stable/server_updates.html#general-update-notes

The point of this would be to have a visual notification of these alerts in the WebUI for dedicated/enroled admins, so that they know they need to intervene ahead of time. 

We already do provide notices about these changes via our release notes, and there are examples you can find related to this in past releases. 

So I hope that makes more sense.

**Comment by Gabgobie** [comment]
> So I hope that makes more sense.

If this is basically about displaying the `FAIL`s of `kanidmd domain upgrade-check` in the future admin ui, yes. Otherwise feel free to take a look at my textified mental tangents ^^

---

<details><summary>My mind going places</summary>

I was aware of the notes you are already putting out which is why I thought it would be a valid fix to suggest new users subscribe to the releases RSS feed.

Having visual admin notifications as part of Kani sounds like an awful lot of effort to solve an issue that I at least to some degree struggle to see. The docs and messages are out here and _could_ be subscribed to without additional maintenance effort and from personal experience, I have never used any account that has administrative privileges in Kani on the web. Maybe this will change with the addition of the admin webui but for now these messages would never reach me.

For the time being, this seems to be blocked by https://github.com/orgs/kanidm/projects/3?pane=issue&itemId=29852289&issue=kanidm%7Ckanidm%7C996 anyways.

---

Given that Kani is not "self-updating" but usually deployed as a container, it's not like we can refuse updating until the admin has addressed all issues. - Well maybe aside from performing migrations and such. This is where I'd currently see the most potential for admin-messaging. We already have `kanidmd domain upgrade-check`. If the admin ignores all documentation, messaging and hasn't [saved/tagged the previous image](https://kanidm.github.io/kanidm/stable/server_updates.html#preserving-the-previous-image), printing out the last working version would help with the rollback.

---

We also can't assume that any given Kani admin is necessarily able to do anything about the migration issue.

> This has happened only once when we changed the format for cron strings, and because the kanidm config is read-only to us, we could not automatically fix this inplace.

This can only be fixed by a someone with filesystem level access. I'd consider this role well above any admin/user tracked in Kani.

---

> This relates to <https://kanidm.github.io/kanidm/stable/server_updates.html#general-update-notes>

> When these are required, **we will give you one release cycle ahead of time to make changes**. To check for changes that will affect your instance you should run.

My understanding of this commitment is as follows:

- The new schema/config/requirement/whatever will be correctly processed in [any(?)](https://kanidm.github.io/kanidm/stable/support.html) version from the previous release cycle.
- Notice about the pending required change/intervention is given in the release notes of versions that are compatible with both variants.

This doesn't seem to be too far off. Going through the releases, the `Important Changes` section seems to announce new requirements.[^1.2][^1.3] Sometimes the changes listed there are not ones that require admin intervention[^1.9], which seems a bit unintuitive to me. Since the linked `1.2/3` changes, the notes got way better in repeating previous patch versions changes in the most recent release.

[^1.2]: 1.2 announcing upcoming changes: <https://github.com/kanidm/kanidm/releases/tag/v1.2.0>
    > In 1.3.0 new constraints will be introduced that limit the gid/uid number ranges for posix accounts. See our [updating documentation](https://kanidm.github.io/kanidm/stable/server_updates.html) which describes steps to detect non-conforming accounts.
[^1.3]: 1.3 making the announced changes a hard requirement: <https://github.com/kanidm/kanidm/releases/tag/v1.3.1>
    > New GID number constraints are now enforced in this version. To upgrade from 1.2.0 all accounts
and groups must adhere to these rules.
[^1.9]: One example for this would be the most recent 1.9 release: <https://github.com/kanidm/kanidm/releases/tag/v1.9.2>

</details>

**Comment by Firstyear** [comment]
> Given that Kani is not "self-updating" but usually deployed as a container, it's not like we can refuse updating until the admin has addressed all issues. - Well maybe aside from performing migrations and such. This is where I'd currently see the most potential for admin-messaging. We already have `kanidmd domain upgrade-check`. If the admin ignores all documentation, messaging and hasn't [saved/tagged the previous image](https://kanidm.github.io/kanidm/stable/server_updates.html#preserving-the-previous-image), printing out the last working version would help with the rollback.
> 
> We also can't assume that any given Kani admin is necessarily able to do anything about the migration issue.
> 
> > This has happened only once when we changed the format for cron strings, and because the kanidm config is read-only to us, we could not automatically fix this inplace.
> 
> This can only be fixed by a someone with filesystem level access. I'd consider this role well above any admin/user tracked in Kani.

Yeah, so this is an argument that we shouldn't pursue this feature at all. The admin needs to be paying attention to the upgrade process. 

We already block the upgrade if it's not satisfied, and we report that to the admin. Generally the only people who get in trouble here are distro package users because distros tend to be very bad about downgrades. 

So perhaps we just consider this as done.

**Comment by Gabgobie** [comment]
> Yeah, so this is an argument that we shouldn't pursue this feature at all. The admin needs to be paying attention to the upgrade process.

Unless anyone can come up with a really good reason for why additional messaging and the associated cost would be necessary.

At the end of the day you can only do so much to prevent the admin from breaking their deployment.

> So perhaps we just consider this as done.

I‘d say so.
